File: 0053-Maximum-Subarray/lc0053.cc
// C++ solution.

// DP.
// Time complexity O(N)
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans = numeric_limits<int>::min();
        int dpval = 0;
        for (vector<int>::const_iterator it = nums.cbegin(); it != nums.cend(); it++) {
            dpval = max(*it + dpval, *it);
            ans = max(dpval, ans);
        }
        return ans;
    }
};

// Greedy.
// Time complexity O(N)
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int subsum = 0;
        int ans = nums[0];
        for (vector<int>::const_iterator it = nums.cbegin(); it != nums.cend(); it++) {
            subsum += *it;
            ans = std::max(ans, subsum);
            if (subsum < 0) { subsum = 0; }
         }
        return ans;
    }
};

File: 0053-Maximum-Subarray/lc0053.rs
// Rust solution.

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut ans = std::i32::MIN;
        if !nums.is_empty() {
            let mut dpval = nums[0];
            ans = nums[0];
            for i in 1..nums.len() {
                dpval = std::cmp::max(nums[i], dpval+nums[i]);
                ans = std::cmp::max(ans, dpval);
            }
        }
        ans
    }
}
